CUDA support in the CSR layout: sparse_to_dense/add_sparse_csr

aocsa opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57403 CUDA support in the CSR layout:  CUDA addmm
* **#57275 CUDA support in the CSR layout: sparse_to_dense/add_sparse_csr**
* #57274 CUDA support in the CSR layout: constructors
closes gh-56485
Differential Revision: [D28112688](https://our.internmc.facebook.com/intern/diff/D28112688)

aocsa committed:
CUDA support in the CSR layout: sparse_to_dense/add_sparse_csr

[ghstack-poisoned]

ezyang commented on aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu:
```diff
@@ -0,0 +1,155 @@
+}
+
+} // namespace native
+} // namespace at
```
lint is failing you on this

aocsa committed:
Update on "CUDA support in the CSR layout: sparse_to_dense/add_sparse_csr"


Differential Revision: [D28112688](https://our.internmc.facebook.com/intern/diff/D28112688)

[ghstack-poisoned]

aocsa committed:
Update on "CUDA support in the CSR layout: sparse_to_dense/add_sparse_csr"

closes gh-56485
Differential Revision: [D28112688](https://our.internmc.facebook.com/intern/diff/D28112688)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28112688).

ezyang commented on aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu:
```diff
@@ -0,0 +1,155 @@
+    const Scalar& alpha) {
+  AT_ASSERT(dense.layout() == kStrided);
+  AT_ASSERT(src.is_sparse_csr());
+  AT_ASSERT(dense.is_cuda());
```
nit TORCH_INTERNAL_ASSERT

ezyang replied:
Also, you'll get sent here even for mixed CPU-CUDA inputs, so modulo @wenleix's patch set about code generating device checks (which I don't think is being applied for add), you DO need to test this explicitly in the body. Don't sweat this too much, a lot of code in PyTorch gets it wrong and generated device checks are the correct fix.

ezyang commented on aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu:
```diff
@@ -0,0 +1,155 @@
+                            auto icol = col_indices_accessor[i];
+                            auto index = out_storage_offset + irow * out_strides0 + icol * out_strides1;
+                            out_ptr[index] += cast_value * values_accessor[i];
+                        }
```
So this got me thinking. COO algorithms are basically CSR algorithms when the COO tensors are coalesced, right? So should we be setting up some code sharing between these two cases (COO coalesced, CSR)?

I went and looked at the COO algorithms; in particular mul (since add takes advantage of the fact that uncoalesced output is OK) and saw this (both sparse inputs were coalesced before this):

```
  Tensor resultNnz = at::empty({1}, CUDA(kLong));
  AT_DISPATCH_ALL_TYPES_AND(
    at::ScalarType::Half, commonDtype, "mul_out_sparse_cuda", [&] {
        apply::valueSparseIntersectionKernel<TensorMulOp<scalar_t>, uint64_t, scalar_t>
          <<<grid, block, 0, stream>>>(
            TensorMulOp<scalar_t>(),
            I_INFO(r_indices_), I_INFO(t_indices_), I_INFO(s_indices_),
            V_INFO(r_values_), V_INFO(t_values_), V_INFO(s_values_),
            static_cast<uint64_t>(t_nnz), static_cast<uint64_t>(s_nnz));
        C10_CUDA_KERNEL_LAUNCH_CHECK();
            
        apply::indexSparseIntersectionKernel<uint64_t, scalar_t>
          <<<1, 1, 0, stream>>>(
            I_INFO(r_indices_), I_INFO(t_indices_), I_INFO(s_indices_),
            // reinterpret_cast shenanigans, because we don't actually have
            // unsigned tensors...
            static_cast<uint64_t>(t_nnz), static_cast<uint64_t>(s_nnz), reinterpret_cast<uint64_t*>(resultNnz.data_ptr()));
        C10_CUDA_KERNEL_LAUNCH_CHECK();
      });   
```

aka we even are launching separate kernels for the values and indices. 

The intersection kernel doesn't look very good though:

```
template <typename Op, typename IndexType, typename Real>
#if __CUDA_ARCH__ >= 350 || defined __HIP_PLATFORM_HCC__
C10_LAUNCH_BOUNDS_2(cuda::getApplyBlockSize(), cuda::getApplyBlocksPerSM())
#endif
__global__ void valueSparseIntersectionKernel(
    Op op,
    TensorInfo<indexT, IndexType> r_indices,
    TensorInfo<indexT, IndexType> t_indices,
    TensorInfo<indexT, IndexType> s_indices,
    TensorInfo<Real, IndexType> r_values,
    TensorInfo<Real, IndexType> t_values,
    TensorInfo<Real, IndexType> s_values,
    const IndexType t_nnz, const IndexType s_nnz) {
  IndexType t_indskip = t_indices.strides[0];
  IndexType s_indskip = s_indices.strides[0];
  int64_t match, d;
  int64_t nDimI = r_indices.sizes[0];
  IndexType valueSize = r_values.strides[0];
  IndexType r_i = 0, t_i = 0, s_i = 0;
  while (t_i < t_nnz && s_i < s_nnz) {
    match = 1;
    for (d = 0; d < nDimI; d++) {
      if (t_indices.data[d * t_indskip + t_i] < s_indices.data[d * s_indskip + s_i]) {
        t_i++;
        match = 0;
        break;
      }
      if (t_indices.data[d * t_indskip + t_i] > s_indices.data[d * s_indskip + s_i]) {
        s_i++;
        match = 0;
        break;
      }
    }
    if (!match) continue;
    applyOp3(op, valueSize, r_values, r_i++, t_values, t_i++, s_values, s_i++);
  }
}
```

so maybe we should replace them with this thrust-based implementation here. Perhaps cc @ngimel

ezyang replied:
You know what, never mind, you need crow indices so doesn't seem like any algorithm sharing can happen

ngimel replied:
We do need to improve sparse multiplication, it's comically bad https://www.internalfb.com/phabricator/paste/view/P140370711. However, due to build and interop issues we are sunsetting thrust usage in the codebase and replacing it with cub wherever possible.

ezyang commented on aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu:
```diff
@@ -0,0 +1,155 @@
+               thrust::for_each(
+                    policy,
+                    thrust::make_counting_iterator(int64_t(0)),
+                    thrust::make_counting_iterator(int64_t(src_crow_indices.size(0) - 1)),
```
NB: as is the classic problem with parallelizing sparse algorithms, this could be wildly imbalanced if the sparsity pattern varies a lot between rows. There isn't an easy fix for this, but it would be nice for the problem to be remarked upon in code.

ezyang commented on aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu:
```diff
@@ -0,0 +1,155 @@
+  Tensor src_crow_indices = src.crow_indices();
+  Tensor src_col_indices = src.col_indices();
+
+  output.resize_as_(dense);
```
use `native::resize_output`here

ezyang commented on aten/src/ATen/native/sparse/cuda/SparseCsrTensorMath.cu:
```diff
@@ -0,0 +1,155 @@
+  Tensor valuesBuffer = src_values.to(commonDtype);
+
+  if (output.scalar_type() != commonDtype) {
+    resultBuffer = dense.to(commonDtype);
```
This looks wrong. If type promotion happens then you will end up writing the result to `resultBuffer` and not whatever the original `output` tensor was. This will concretely manifest with something like `torch.add(dense_float, sparse_float, out=dense_double)`

ezyang commented:
The testing in this PR is woefully insufficient.

ezyang commented:
It is nice seeing this PR make an attempt at getting type promotion correct, even though it doesn't actually manage to do so. This is one of those things where I don't think you'll be able to fumble your way into correctness, so we should either (1) agree not to make an attempt at doing it properly, or (2) setup a call so we can talk through the high level algorithmics of how type promotion works (maybe this would be a good topic for my up and coming podcast). LMK.

aocsa committed:
Update on "CUDA support in the CSR layout: sparse_to_dense/add_sparse_csr"

closes gh-56485
Differential Revision: [D28112688](https://our.internmc.facebook.com/intern/diff/D28112688)

[ghstack-poisoned]

aocsa committed:
Update on "CUDA support in the CSR layout: sparse_to_dense/add_sparse_csr"

closes gh-56485
Differential Revision: [D28112688](https://our.internmc.facebook.com/intern/diff/D28112688)

[ghstack-poisoned]

aocsa commented:
> It is nice seeing this PR make an attempt at getting type promotion correct, even though it doesn't actually manage to do so. This is one of those things where I don't think you'll be able to fumble your way into correctness, so we should either (1) agree not to make an attempt at doing it properly, or (2) setup a call so we can talk through the high level algorithmics of how type promotion works (maybe this would be a good topic for my up and coming podcast). LMK.

Thanks, @ezyang, for the review. I rebased this PR on top of the latest changes, polished a few things based on the feedback, fixed the type promotion for CPU/CUDA and added new tests for `torch.add(dense, sparse_csr)`.  I will add `torch.add(sparse_csr, sparse_csr)` op support for CPU/CUDA  in a new PR in this stack.  cc @pearu

aocsa committed:
Update on "CUDA support in the CSR layout: sparse_to_dense/add_sparse_csr"

closes gh-56485
Differential Revision: [D28112688](https://our.internmc.facebook.com/intern/diff/D28112688)

[ghstack-poisoned]

aocsa committed:
Update on "CUDA support in the CSR layout: sparse_to_dense/add_sparse_csr"

closes gh-56485
Differential Revision: [D28112688](https://our.internmc.facebook.com/intern/diff/D28112688)

[ghstack-poisoned]

aocsa committed:
Update on "CUDA support in the CSR layout: sparse_to_dense/add_sparse_csr"

closes gh-56485
Differential Revision: [D28112688](https://our.internmc.facebook.com/intern/diff/D28112688)

[ghstack-poisoned]

aocsa committed:
Update on "CUDA support in the CSR layout: sparse_to_dense/add_sparse_csr"

closes gh-56485
Differential Revision: [D28112688](https://our.internmc.facebook.com/intern/diff/D28112688)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28112688).

aocsa committed:
addressed review comments on "CUDA support in the CSR layout: sparse_to_dense/add_sparse_csr"

closes gh-56485
Differential Revision: [D28112688](https://our.internmc.facebook.com/intern/diff/D28112688)

[ghstack-poisoned]

aocsa committed:
review comments addressed on "CUDA support in the CSR layout: sparse_to_dense/add_sparse_csr"

closes gh-56485
Differential Revision: [D28112688](https://our.internmc.facebook.com/intern/diff/D28112688)

[ghstack-poisoned]

aocsa committed:
minor lint fix on "CUDA support in the CSR layout: sparse_to_dense/add_sparse_csr"

closes gh-56485
Differential Revision: [D28112688](https://our.internmc.facebook.com/intern/diff/D28112688)

[ghstack-poisoned]

pearu reviewed (approved):
LGTM

aocsa committed:
addressed review comments - 2 on "CUDA support in the CSR layout: sparse_to_dense/add_sparse_csr"

closes gh-56485
Differential Revision: [D28112688](https://our.internmc.facebook.com/intern/diff/D28112688)

[ghstack-poisoned]

aocsa committed:
rebased/addressed review comments - 3 on "CUDA support in the CSR layout: sparse_to_dense/add_sparse_csr"

closes gh-56485
Differential Revision: [D28112688](https://our.internmc.facebook.com/intern/diff/D28112688)

[ghstack-poisoned]

aocsa committed:
rebased/last feedback addressed on "CUDA support in the CSR layout: sparse_to_dense/add_sparse_csr"

closes gh-56485
Differential Revision: [D28112688](https://our.internmc.facebook.com/intern/diff/D28112688)

[ghstack-poisoned]

aocsa committed:
run CI/last feedback addressed on "CUDA support in the CSR layout: sparse_to_dense/add_sparse_csr"

closes gh-56485
Differential Revision: [D28112688](https://our.internmc.facebook.com/intern/diff/D28112688)

[ghstack-poisoned]

aocsa committed:
rebased needed on "CUDA support in the CSR layout: sparse_to_dense/add_sparse_csr"

closes gh-56485
Differential Revision: [D28112688](https://our.internmc.facebook.com/intern/diff/D28112688)

[ghstack-poisoned]

aocsa committed:
rebased needed : on "CUDA support in the CSR layout: sparse_to_dense/add_sparse_csr"

closes gh-56485
Differential Revision: [D28112688](https://our.internmc.facebook.com/intern/diff/D28112688)

[ghstack-poisoned]

aocsa committed:
rebased needed after csr addmm fix is merged on "CUDA support in the CSR layout: sparse_to_dense/add_sparse_csr"

closes gh-56485
Differential Revision: [D28112688](https://our.internmc.facebook.com/intern/diff/D28112688)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28112688).

aocsa committed:
minor fix for to_dense CUDA/HIP test error on "CUDA support in the CSR layout: sparse_to_dense/add_sparse_csr"

closes gh-56485
Differential Revision: [D28112688](https://our.internmc.facebook.com/intern/diff/D28112688)

[ghstack-poisoned]

aocsa committed:
minor fix for to_dense CUDA/HIP test error on "CUDA support in the CSR layout: sparse_to_dense/add_sparse_csr"

closes gh-56485
Differential Revision: [D28112688](https://our.internmc.facebook.com/intern/diff/D28112688)

[ghstack-poisoned]

aocsa committed:
minor update to PR 57275 on "CUDA support in the CSR layout: sparse_to_dense/add_sparse_csr"

closes gh-56485
Differential Revision: [D28112688](https://our.internmc.facebook.com/intern/diff/D28112688)

[ghstack-poisoned]

aocsa commented:
This PR and the next in the stack is also ready to be merge. pytorch_linux_backward_compatibility_check_test failed but this is not related to changes in this PR, pretty sure that it only need to rebase. Let me know is if it's needed or not to re-import again. Thanks @ezyang, @malfet, @bhosmer.

aocsa commented:
Closing this PR too :(. Stack PR was poisoned after first  PR in the stack was reverted. New PR is here:
https://github.com/pytorch/pytorch/pull/59004

bhosmer commented:
new new PR: #59011 (thanks for restacking @aocsa)
